### PR TITLE
Added a self-closing iterator class.

### PR DESCRIPTION
**Pull request by tfenne**

### CODE REVIEW

File: src/main/scala/com/fulcrumgenomics/commons/collection/SelfClosingIterator.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    */
+  override def next(): A = {
+    val result = super.next()
+    if (!hasNext) close()
+    result
+  }
+
+  /** Invokes the closer function provided in the constructor. */
+  override def close(): Unit = if (open) {
+    open = true
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/commons/collection/SelfClosingIterator.scala]: Uh, shouldn't this be `open = false`?  Could you add a test that you check fails before you change it?

File: src/test/scala/com/fulcrumgenomics/commons/collection/SelfClosingIteratorTest.scala [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package com.fulcrumgenomics.commons.collection
+
+import com.fulcrumgenomics.commons.util.UnitSpec
+
+class SelfClosingIteratorTest extends UnitSpec {
+  "SelfClosingIterator" should "return all the items and then call the close function" in {
```
Comment (nh13) [inline src/test/scala/com/fulcrumgenomics/commons/collection/SelfClosingIteratorTest.scala:6]: This test wont check that `() => closed` isn't called twice.  A better test would be to have
`var closed = 0` and `() => closed += 1`, and then check the value of `closed` (should be `1`).

**Review by nh13 — CHANGES_REQUESTED** [review]
A little but important bug to fix.

File: src/test/scala/com/fulcrumgenomics/commons/collection/SelfClosingIteratorTest.scala (hunk 2 above)
Comment (tfenne) [reply src/test/scala/com/fulcrumgenomics/commons/collection/SelfClosingIteratorTest.scala:6]: Added another test.

File: src/main/scala/com/fulcrumgenomics/commons/collection/SelfClosingIterator.scala (hunk 1 above)
Comment (tfenne) [reply src/main/scala/com/fulcrumgenomics/commons/collection/SelfClosingIterator.scala]: Doh.  I changed the var from `closed` to `open` and apparently missed this.  Thanks for the catch.

**Comment by tfenne** [comment]
All fixed up and two more tests.  Want to take a second look @nh13?

**Review by nh13 — APPROVED** [review]
LGTM

**Pull request merged**